doc/doxygen/src/build-system-cpu-conventions.md

```diff
@@ -0,0 +1,177 @@
+good candidates.
+
+The main goal is to remove the algorithmic and shell based resolutions for this
+if possible.
+
+### File based handling ###
+
+Somehow, having all `CPU_MODEL` in one directory is just to save common
+parts and keep the cpu directory tidy, all the different ones could still
+be defined in there own directory, but nested in the common one.
+
+This could transfer to have for each `CPU_MODEL` for a `CPU` the following files:
+
+~~~~~~~~~~~~~~~~~~~ {.mk}
+$(RIOTCPU)/$(CPU)/models/$(CPU_MODEL)/Makefile.features
+$(RIOTCPU)/$(CPU)/models/$(CPU_MODEL)/Makefile.dep
+$(RIOTCPU)/$(CPU)/models/$(CPU_MODEL)/Makefile.include
+$(RIOTCPU)/$(CPU)/models/$(CPU_MODEL)/doc.txt
+~~~~~~~~~~~~~~~~~~~
+
+And then each main `$(CPU)` file will include the `$(CPU_MODEL)` specific one.
+
+This would match the way to do it in RIOT but could create many
+directories and files for maybe no or only few lines.
+
+It could make sense to go with this if the files, especially the
+`Makefile.dep` or `Makefile.include` ones would need many specific definitions
+and handling that are more than variables configuration for the main `CPU`
+files.
+
+
+### `ifneq` based handling ###
+
+To not add any files, the main `CPU` files could add specific definitions if the
+`CPU_MODEL` matches certain values.
+
+~~~~~~~~~~~~~~~~~~~ {.mk}
+ifneq (,$(filter abc1 abd0 ade5,$(CPU_MODEL)))
+  FEATURES_PROVIDED += periph_hwrng
+endif
+~~~~~~~~~~~~~~~~~~~
+
+This allow simple handling when the benefit of having per model definition
+would currently not overcome the overhead of defining it.
+Also, as currently dependencies are not declarative, it is with the previous
+solution one of the only alternative as the following declarative only would
+not work.
+
+An issue with this is that this it not declarative and does not have equivalent
+in a declarative only language that is not Turing complete as `make`.
+But as long as it used with explicit inclusion lists (no wildcard) and no
+blacklists, it is easy to map it to a per model definition from the other
+methods.
+
+
+### Exhaustive per model definition ###
+
+In the case of only declarative values of specific variables, handling different
+`CPU_MODEL` is simply defining one per `CPU_MODEL`.
+When not relying on an `if` evaluation as before, one could define dictionaries
+which are implemented in `make` when using only one file with "namespaced"
+variables names including the `key` and so the `CPU_MODEL` value in the name.
+
+~~~~~~~~~~~~~~~~~~~ {.mk}
+ROM_LEN_stm32f103c8 = 64K
+ROM_LEN_stm32f103cb = 128K
+ROM_LEN_stm32f103rb = 128K
+ROM_LEN_stm32f103re = 512K
+
+ROM_LEN = ROM_LEN_$(CPU_MODEL)
+~~~~~~~~~~~~~~~~~~~
+
+This represents a direct mapping of the `CPU_MODEL` to the variable value.
+This exact format makes sense when there are not too many variables or for
+variables that have a list as value as the following solution does not apply.
+This is true for example for `FEATURES_PROVIDED`.
```

Comment:
or for variables whose value is a list of values (this is case where solution 4 does not apply). This is true ...

Reply:
Yep better